Index: HierarchicalSearch.c
===================================================================
RCS file: /usr/local/cvs/lscsoft/lalapps/src/pulsar/hough/src2/HierarchicalSearch.c,v
retrieving revision 1.166
retrieving revision 1.167
diff -u -r1.166 -r1.167
--- HierarchicalSearch.c	14 May 2007 11:56:08 -0000	1.166
+++ HierarchicalSearch.c	5 Jun 2007 15:42:30 -0000	1.167
@@ -327,6 +327,7 @@
   DopplerSkyScanState thisScan = empty_DopplerSkyScanState; /* current state of the Doppler-scan */
   static PulsarDopplerParams dopplerpos;	       /* current search-parameters */
   static PulsarDopplerParams thisPoint; 
+  REAL8  initialPatchSizeY;
 
   /* temporary storage for spinrange vector */
   static PulsarSpinRange spinRange_Temp;
@@ -848,6 +849,8 @@
     semiCohPar.patchSizeY = 1.0 / ( tStack * usefulParams.spinRange_midTime.fkdot[0] * VEPI ); 
   }
     
+  initialPatchSizeY = semiCohPar.patchSizeY;
+  
   LogPrintf(LOG_DETAIL,"Hough/Stackslide patchsize is %frad x %frad\n", 
 	    semiCohPar.patchSizeX, semiCohPar.patchSizeY); 
 
@@ -926,6 +929,17 @@
       semiCohPar.alpha = thisPoint.Alpha;
       semiCohPar.delta = thisPoint.Delta;
       
+      semiCohPar.patchSizeY = initialPatchSizeY;
+      
+      if( (semiCohPar.delta > 0) && (semiCohPar.delta< 0.5*(LAL_PI -semiCohPar.patchSizeX) ) )
+         semiCohPar.patchSizeY = initialPatchSizeY*cos(semiCohPar.delta - 0.5*semiCohPar.patchSizeX)/cos(semiCohPar.delta);
+
+      if( (semiCohPar.delta < 0) && (semiCohPar.delta> 0.5*(-LAL_PI +semiCohPar.patchSizeX) ) )
+         semiCohPar.patchSizeY = initialPatchSizeY*cos(semiCohPar.delta + 0.5*semiCohPar.patchSizeX)/cos(semiCohPar.delta);
+
+      /* assuming semiCohPar.delta \in [-pi/2, pi/2], otherwise the conditions should be more robust */
+
+      
       /* initialize weights to unity */
       LAL_CALL( LALHOUGHInitializeWeights( &status, weightsV), &status);
 
